kani-compiler/src/kani_middle/attributes.rs

```diff
@@ -2,33 +2,278 @@
+        tcx.sess.span_err(
+            attr.span,
+            &format!("only one '#[kani::{}]' attribute is allowed per harness", kind.as_ref()),
+        );
+    }
+    attr
+}
+
+/// Check that if an item is tagged with a proof_attribute, it is a valid harness.
+fn check_proof_attribute(tcx: TyCtxt, def_id: DefId, proof_attributes: &Vec<&Attribute>) {
+    assert!(!proof_attributes.is_empty());
+    let span = proof_attributes.first().unwrap().span;
+    if proof_attributes.len() > 1 {
+        tcx.sess.span_warn(proof_attributes[0].span, "duplicate attribute");
+    }
+
+    if tcx.def_kind(def_id) != DefKind::Fn {
+        tcx.sess.span_err(span, "the `proof` attribute can only be applied to functions");
+    } else if tcx.generics_of(def_id).requires_monomorphization(tcx) {
+        tcx.sess.span_err(span, "the `proof` attribute cannot be applied to generic functions");
+    } else {
+        let instance = Instance::mono(tcx, def_id);
+        let helper = CompilerHelpers { tcx };
+        if !helper.fn_abi_of_instance(instance, ty::List::empty()).args.is_empty() {
+            tcx.sess.span_err(span, "functions used as harnesses cannot have any arguments");
+        }
+    }
+}
+
+/// Partition all the attributes according to their kind.
+fn extract_kani_attributes(
+    tcx: TyCtxt,
+    def_id: DefId,
+) -> BTreeMap<KaniAttributeKind, Vec<&Attribute>> {
+    let all_attributes = tcx.get_attrs_unchecked(def_id);
+    all_attributes.iter().fold(BTreeMap::default(), |mut result, attribute| {
         // Get the string the appears after "kanitool::" in each attribute string.
         // Ex - "proof" | "unwind" etc.
-        if let Some(attribute_string) = kanitool_attr_name(attr).as_deref() {
-            if attribute_string == "proof" {
-                proof_attributes.push(attr);
+        if let Some(kind) = attr_kind(tcx, attribute) {
+            result.entry(kind).or_default().push(attribute)
+        }
+        result
+    })
+}
+
+/// Return the unwind value from the given attribute.
+fn parse_unwind(tcx: TyCtxt, attr: &Attribute) -> Option<u32> {
+    // Get Attribute value and if it's not none, assign it to the metadata
+    match parse_integer(attr) {
+        None => {
+            // There are no integers or too many arguments given to the attribute
+            tcx.sess.span_err(
+                attr.span,
+                "invalid argument for `unwind` attribute, expected an integer",
+            );
+            None
+        }
+        Some(unwind_integer_value) => {
+            if let Ok(val) = unwind_integer_value.try_into() {
+                Some(val)
             } else {
-                other_attributes.push((attribute_string.to_string(), attr));
+                tcx.sess.span_err(attr.span, "value above maximum permitted value - u32::MAX");
+                None
             }
         }
     }
+}
 
-    (proof_attributes, other_attributes)
+fn parse_stubs(tcx: TyCtxt, attributes: Vec<&Attribute>) -> Vec<Stub> {
```

Comment:
I'm assuming this function ensures the right format for the `kani::Stub` attribute? Could you add a comment above the function?

Reply:
This is converting the attributes into a list of stubs to be applied. Yes, I can totally add a comment!